src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,171 @@ namespace scene {
+
+  MagnumObject::setParent(newParent);
+
+  // Update new ancestors'SubtreeSensorSuites
+  addSubtreeSensorsToAncestors();
+  // Add sensors to newParent's nodeSensorSuite
+  addSensorToParentNodeSensorSuite();
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, add it to parent's nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
+                   "SceneNode::addSensorToParentNodeSensorSuite(): Parent null "
+                   "is nullptr", );
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .add(it->second);
+  }
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only remove a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  // Get the first one, and if it is valid, remove it from parent's
+  // nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    CORRADE_ASSERT(static_cast<scene::SceneNode*>(this->parent()) != nullptr,
+                   "SceneNode::addSensorToParentNodeSensorSuite(): Parent null "
+                   "is nullptr", );
+    static_cast<scene::SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .remove(it->first);
+  }
+}
+
+void SceneNode::addSubtreeSensorsToAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
+      currentNode->getSubtreeSensorSuite().add(sensor.second);
+      currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
+    }
+  }
+}
+
+void SceneNode::removeSubtreeSensorsFromAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
+      currentNode->getSubtreeSensorSuite().remove(sensor.first);
+      currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
```

Comment:
Again, what if the currentNode becomes `nullptr`? It can happen during the deconstruction process.

Reply:
We check that currentNode is not nullptr in the while loop condition.